python/ray/serve/tests/test_backend_state.py

```diff
@@ -1,89 +1,258 @@
-import pytest
-from typing import Optional, Tuple
+import time
+from typing import Dict, Optional, Tuple
 from unittest.mock import patch, Mock
-from uuid import uuid4
 
-from ray.serve.common import BackendConfig, BackendInfo, ReplicaConfig
-from ray.serve.backend_state import BackendState
+import pytest
+
+from ray.actor import ActorHandle
+from ray.serve.async_goal_manager import AsyncGoalManager
+from ray.serve.common import (
+    BackendConfig,
+    BackendInfo,
+    BackendTag,
+    ReplicaConfig,
+    ReplicaTag,
+)
+from ray.serve.backend_state import BackendState, ReplicaState
+
+
+def mock_replica_factory(mock_replicas):
+    class MockReplicaActorWrapper:
+        def __init__(self, actor_name: str, detached: bool,
+                     controller_name: str, replica_tag: ReplicaTag,
+                     backend_tag: BackendTag):
+            self._actor_name = actor_name
+            self._replica_tag = replica_tag
+            self._backend_tag = backend_tag
+            self._state = ReplicaState.SHOULD_START
+
+            mock_replicas.append(self)
+
+            # Will be set when `start()` is called.
+            self.started = False
+            # Expected to be set in the test.
+            self.ready = False
+            # Will be set when `graceful_stop()` is called.
+            self.stopped = False
```

Comment:
nit: could this be something like `graceful_stop`?

Reply:
yep will doa

Reply:
actually this is going to cause me some annoying merge conflicts so i'll change it in another PR!